bpt_writer_frames.cpp

```diff
@@ -0,0 +1,289 @@
+            if (f.effect() & WR) {
+                bytes_type::value_type b(f.value(e.bytes()));
+                add(W, f.name(), bytes_type(1, b), f.width());
+            }
+        }
+    }
+
+    ~std_frame_element() {
+        cont.add(frm);
+    }
+
+private:
+    void add(usage u, const register_event& e) {
+        add(u, e.name(), e.bytes(), e.width()); 
+    }
+
+    void add(usage u, const memory_event& e) {
+        add(u, e.addr(), e.bytes());
+    }
+
+    void add(usage u, ADDRINT addr, const bytes_type& data) {
+        ::operand_info_specific* ois = create_operand(u, data);
+        ::mem_operand* mo = ois->mutable_mem_operand();
+        mo->set_address(addr);
+    }
+
+    void add(usage u, const std::string& name,
+                     const bytes_type& data, int width) {
+        ::operand_info_specific* ois = create_operand(u, data, width);
+        ::reg_operand* ro = ois->mutable_reg_operand();
+        ro->set_name(name);
+    }
+
+
+    ::operand_info_specific* create_operand(usage u,
+                                            const bytes_type& data,
+                                            int width = 0) {
+        ::std_frame* sf = frm.mutable_std_frame();
+        ::operand_value_list* ovl = 0;
+        switch(u) {
+        case R: ovl = sf->mutable_operand_pre_list(); break;
+        case W: ovl = sf->mutable_operand_post_list(); break;
+        }
+        ::operand_info* oi = ovl->add_elem();
+        oi->set_bit_length(width == 0 ? 8*data.size() : width);
+        ::operand_usage* ou = oi->mutable_operand_usage();
+        ou->set_read(u == R);
+        ou->set_written(u == W);
+        ou->set_index(false);
+        ou->set_base(false);
+        oi->set_value(std::string(data.begin(), data.end()));
+        oi->mutable_taint_info()->set_no_taint(true);
+        return oi->mutable_operand_info_specific();
+    }
+
+private:
+    container_type& cont;
+    ::frame frm;
+};
+
+static const frame_architecture arch = frame_arch_i386;
+#if defined(TARGET_IA32)
+    static const uint64_t machine = frame_mach_i386_i386;
+#elif defined(TARGET_IA32E)
+    static const uint64_t machine = frame_mach_x86_64;
+#else
+#error "Usupported machine"
+#endif
+
+namespace meta {
+template <typename T>
+void init_envp(T* record, char* envp[]) {
+    while(*envp) {
+        record->add_envp(*envp);
+        ++envp;
+    }
+}
+
+template <typename T>
+void init_args(T* record, int argc, char* argv[]) {
+    for (int i = 0; i < argc; ++i) {
+        record->add_args(argv[i]);
+    }
+}
+
+void init_tracer(::tracer* tracer,
+                 int argc, char* argv[], char* envp[]) {
+    tracer->set_name("bpt");
+    init_args(tracer, argc, argv);
+    init_envp(tracer, envp);
+    tracer->set_version("1.0.0/bpt");
+}
+
+std::string md5sum(const std::string& path) {
+    std::string md5;
+    CryptoPP::Weak::MD5 hash;
+    CryptoPP::FileSource(
+        path.c_str(), true,
+        new CryptoPP::HashFilter(
```

Comment:
[FileSource](https://cryptopp.com/wiki/FileSource)  deletes pointers by itself (noted in second example)